Web Platform Tests for CSS shorthands have a few mistakes I noticed when fixing shorthand bugs

darinadler opened the pull request:
#### d8d729c265469a5a48003a3e972b986e1dfa6e8b
<pre>
Web Platform Tests for CSS shorthands have a few mistakes I noticed when fixing shorthand bugs
<a href="https://bugs.webkit.org/show_bug.cgi?id=249388">https://bugs.webkit.org/show_bug.cgi?id=249388</a>
rdar://problem/103399812

Reviewed by Tim Nguyen.

These tests are corrected to expect the correct behavior. The expected results, however, still
reflect the behavior of the current WebKit code. A future pull request will progress these tests.
I came up with these changes by testing with those improvements and studying what these tests
expect, and changing the tests in the cases where it was the tests that were wrong.

* LayoutTests/cssom/cssvalue-comparison-expected.txt: Updated.
* LayoutTests/cssom/cssvalue-comparison.html: Reworked test so it no longer depends on reading
back shorthands, since that gives false negatives.

* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize-expected.txt: Updated.
* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize.html: Expect just
border-box instead of border-box border-box.

* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt: Updated.
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style.html: Expect initial
values to be omitted. Also updated one case that expected the initial identifier but should get the
initial value instead.

* LayoutTests/fast/css/background-clip-text-expected.txt: Updated.
* LayoutTests/fast/css/background-clip-text.html: Expect border-box instead of initial.

* LayoutTests/fast/css/border-shorthand-initialize-longhands-expected.txt: Updated.
* LayoutTests/fast/css/border-shorthand-initialize-longhands.html: Expect initial values instead of
the initial identifier.

* LayoutTests/fast/css/cssText-shorthand-expected.txt: Updated.
* LayoutTests/fast/css/cssText-shorthand.html: Removed most of what is tested here since this test
was mostly imported into Web Platform Tests long ago. All that remains are WebKit-specific bits.

* LayoutTests/fast/css/font-shorthand-from-longhands-expected.txt: Updated.
* LayoutTests/fast/css/font-shorthand-from-longhands.html: Expect correct behavior from font
shorthand where it will not serialize unless all the longhands are set.

* LayoutTests/fast/css/getPropertyValue-border.html: Expect initial value instead of the initial
identifier.

* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing-expected.txt: Updated.
* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing.html: Expect initial values to be
omitted from the animation shorthand.

* LayoutTests/fast/css/parsing-text-emphasis-expected.txt: Updated.
* LayoutTests/fast/css/parsing-text-emphasis.html: Expect initial value instead of the initial
identifier.

* LayoutTests/fast/css/shorthand-mismatched-list-crash-expected.txt: Updated.
* LayoutTests/fast/css/shorthand-mismatched-list-crash.html: Expect initial values to be omitted
from the -webkit-transition shorthand.

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt: Updated.
* LayoutTests/fast/dom/background-shorthand-csstext.html: Updated test so we can see the results
rather than just that there was a failure. Updated expectations to not expect serialized none and
scroll, which are the initial values.

* LayoutTests/fast/masking/parsing-mask-expected.txt: Updated.
* LayoutTests/fast/masking/parsing-mask.html: Expect initial values to be omitted from shorthands.

* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt: Updated.
* LayoutTests/fast/masking/parsing-webkit-mask.html: Expect initial values to be omitted from
shorthands.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-valid.html:
Expect either 0s or none for animation shorthand when everything is set to the default, since
animation-duration is the first longhand. Allow initial values to be omitted in the layered
test case at the end of this test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt:
Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid.html:
Expect center to be omitted when it can be without changing the meaning.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html: Expect
to correctly not serialize none for the mask-reference since it&apos;s the initial value, and some
other cases here that had incorrect expectations.

* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText-expected.txt:
Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText.html: Expect a
single space before &quot;!important&quot;. The specification explicitly calls for a single space, not two
spaces, so I didn&apos;t bother to tolerate both formats.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt: Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values.html: Added an exceptions
list for all the cases where &quot;center&quot; must be omitted for minimal serialization.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt: Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values.html: Corrected
expectations. This test was expecting incorrect behavior for some tricky cases, including some
that are out of date since the specifiation now says that the border shorthand sets all the
longhands of the the border-image shorthand to their initial values.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Updated.

Canonical link: <a href="https://commits.webkit.org/257966@main">https://commits.webkit.org/257966@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a93024b052ac7a87ff8c49b5c04d5011f8c4bc28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100528 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33571 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/109835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10603 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/107690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106307 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3401 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9516 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/43666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5456 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/5202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/cssom/cssvalue-comparison.html:
```diff
@@ -10,71 +10,88 @@
+        styleTextNode.data = "div { " + propertyName + " : " + styleSheetPropertyValue + "; }"
+        document.execCommand('insertHTML', false, "<div id=\"insertedDiv\" style=\"" + propertyName + " : " + inlineStylePropertyValue + "; \"></div>")
+        var insertedDiv = document.getElementById('insertedDiv')
+        var result = insertedDiv.style[propertyName] === ""
```
Is this useful? `result` is overridden below.

darinadler replied:
I will check.

darinadler replied:
Yes, left it in by accident. Deleted now.

nt1m commented on LayoutTests/cssom/cssvalue-comparison.html:
```diff
@@ -10,71 +10,88 @@
-        var testMap = tests[index];
-        for (var key in testMap) {
-            var testValues = testMap[key];
+   for (var i in tests) {
```
`for (let [propertyName, testValues] of tests) {` would be more readable (and would make the code below more simple too.

darinadler replied:
OK

darinadler replied:
Done.

nt1m commented on LayoutTests/cssom/cssvalue-comparison.html:
```diff
@@ -10,71 +10,88 @@
-            for(var testIndex in testValues)
-                test(function() {assert_true(runTest(key, testValues[testIndex], testValues[testIndex], true))}, "Two CSSValues \"" + testValues[testIndex] + "\" for property \"" + key + "\" are equal.");
+        for (var j in testValues) {
+            var value = testValues[j]
```
`for (const value of testValues)`

darinadler replied:
All right.

nt1m commented on LayoutTests/fast/css/cssText-shorthand.html:
```diff
@@ -1,41 +1,17 @@
     ['-webkit-border-horizontal-spacing: 1px; -webkit-border-vertical-spacing: 2px;', 'border-spacing: 1px 2px;'],
 
-    // We don't use shorthand for font-family, etc... for compatibility reasons
+    // We don't use the font shorthand for this case yet, for historical editing compatibility reasons.
```
We probably should at some point fwiw. The editing code will need work though.

darinadler replied:
Yes I agree completely and would like to do this work myself if I can. I will try it after finishing with implicit.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-valid.html:
```diff
@@ -22,14 +22,14 @@
 test_valid_value("animation", "both", ["both", "0s ease 0s 1 normal both running none"]);
 test_valid_value("animation", "paused", ["paused", "0s ease 0s 1 normal none paused none"]);
-test_valid_value("animation", "none", ["none", "0s ease 0s 1 normal none running none"]);
+test_valid_value("animation", "none", ["0s", "none", "0s ease 0s 1 normal none running none"]);
```
Perhaps a personal preference, but I have a bad feeling about using `0s` as serialization here. I could see web content checking that `none` is contained in the serialization.

darinadler replied:
I am 100% fine sticking with `none` here, but what is the rule that follows? Would you like me to just code an exception for this case or do you have a rule to propose?

The rule I was following was to serialize the initial value of the first longhand when all longhands are equal to their initial values.

nt1m replied:
https://www.w3.org/TR/cssom-1/#serializing-css-values
> If either of the above syntactic translations would be less backwards-compatible, do not perform them.

I admit I haven't done any real research on that bit, but I wouldn't be surprised if a regression came up because some animation library or website searched for `none` in the serialization. @karlcow Do you know if serializing to `animation: 0s` instead of `animation: none` could cause an issue ?

Anyway, I'm not blocking on this, but something to think about.

darinadler replied:
I would like to choose whether to switch to `0s` or not. I am happy to code it as a special case for website compatibility, real or imagined, or to invent some rule (not sure what the rule would be). I understand your compatibility worry, but I will note that many of the CSS serialization changes we are making come with compatibility risk, and to me this particular one doesn’t seem greater than the others.

karlcow replied:
Are we talking about this case?

```
data:text/html,<div%20style="animation:none">
```

So right now.
WebKit, Gecko and Blink all return:

```
0s ease 0s 1 normal none running none
```

I checked a bit. I don't know about any issues. Specifically I see people assigning the value to none, but I didn't find **yet**, people testing against  `none` for the matter. 

Example: 

https://github.com/ashwinexe/Prehackstoric-Hack/blob/5bcd76f39489a450d12f3b36385aa2df76b8573f/scripts/game.js#L70

0 occurrence of `none ease 0s 1 normal none running none`
272 occurrences of `0s ease 0s 1 normal none running none`

Assuming I got the framing right at the start of this comment.

darinadler replied:
Thanks Karl, you got it exactly right.

nt1m replied:
@karlcow I was wondering if they do checks like `style.animation.includes("none")` or such. I couldn't find anything from a quick Github search, so we might be safe here :) Checks like the one you found are also problematic, but we might just have to deal with it.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values.html:
```diff
@@ -140,11 +140,33 @@
+    };
+
+    function create_result(propertyName, actual, expected) {
+      var key = propertyName + ": " + expected
+      if (key in minimal_results)
+        expected = minimal_results[key]
```
Is this the same as:

```
if (propertyName == "background-position")
  expected = expected.replace("center", "").trim();
```

darinadler replied:
No.

When both are identifiers we can omit center if it's first or second.

When one is an identifier and the other is a number we can omit center if it's second but not if it's first.

When one is a pair then we can't omit center even if it's after the pair.

nt1m replied:
Ah right, makes sense.

darinadler committed:
Web Platform Tests for CSS shorthands have a few mistakes I noticed when fixing shorthand bugs
https://bugs.webkit.org/show_bug.cgi?id=249388
rdar://problem/103399812

Reviewed by Tim Nguyen.

These tests are corrected to expect the correct behavior. The expected results, however, still
reflect the behavior of the current WebKit code. A future pull request will progress these tests.
I came up with these changes by testing with those improvements and studying what these tests
expect, and changing the tests in the cases where it was the tests that were wrong.

* LayoutTests/cssom/cssvalue-comparison-expected.txt: Updated.
* LayoutTests/cssom/cssvalue-comparison.html: Reworked test so it no longer depends on reading
back shorthands, since that gives false negatives.

* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize-expected.txt: Updated.
* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize.html: Expect just
border-box instead of border-box border-box.

* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt: Updated.
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style.html: Expect initial
values to be omitted. Also updated one case that expected the initial identifier but should get the
initial value instead.

* LayoutTests/fast/css/background-clip-text-expected.txt: Updated.
* LayoutTests/fast/css/background-clip-text.html: Expect border-box instead of initial.

* LayoutTests/fast/css/border-shorthand-initialize-longhands-expected.txt: Updated.
* LayoutTests/fast/css/border-shorthand-initialize-longhands.html: Expect initial values instead of
the initial identifier.

* LayoutTests/fast/css/cssText-shorthand-expected.txt: Updated.
* LayoutTests/fast/css/cssText-shorthand.html: Removed most of what is tested here since this test
was mostly imported into Web Platform Tests long ago. All that remains are WebKit-specific bits.

* LayoutTests/fast/css/font-shorthand-from-longhands-expected.txt: Updated.
* LayoutTests/fast/css/font-shorthand-from-longhands.html: Expect correct behavior from font
shorthand where it will not serialize unless all the longhands are set.

* LayoutTests/fast/css/getPropertyValue-border.html: Expect initial value instead of the initial
identifier.

* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing-expected.txt: Updated.
* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing.html: Expect initial values to be
omitted from the animation shorthand.

* LayoutTests/fast/css/parsing-text-emphasis-expected.txt: Updated.
* LayoutTests/fast/css/parsing-text-emphasis.html: Expect initial value instead of the initial
identifier.

* LayoutTests/fast/css/shorthand-mismatched-list-crash-expected.txt: Updated.
* LayoutTests/fast/css/shorthand-mismatched-list-crash.html: Expect initial values to be omitted
from the -webkit-transition shorthand.

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt: Updated.
* LayoutTests/fast/dom/background-shorthand-csstext.html: Updated test so we can see the results
rather than just that there was a failure. Updated expectations to not expect serialized none and
scroll, which are the initial values.

* LayoutTests/fast/masking/parsing-mask-expected.txt: Updated.
* LayoutTests/fast/masking/parsing-mask.html: Expect initial values to be omitted from shorthands.

* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt: Updated.
* LayoutTests/fast/masking/parsing-webkit-mask.html: Expect initial values to be omitted from
shorthands.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-valid.html:
Expect either 0s or none for animation shorthand when everything is set to the default, since
animation-duration is the first longhand. Allow initial values to be omitted in the layered
test case at the end of this test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt:
Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid.html:
Expect center to be omitted when it can be without changing the meaning.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html: Expect
to correctly not serialize none for the mask-reference since it's the initial value, and some
other cases here that had incorrect expectations.

* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText-expected.txt:
Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText.html: Expect a
single space before "!important". The specification explicitly calls for a single space, not two
spaces, so I didn't bother to tolerate both formats.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt: Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values.html: Added an exceptions
list for all the cases where "center" must be omitted for minimal serialization.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt: Updated.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values.html: Corrected
expectations. This test was expecting incorrect behavior for some tricky cases, including some
that are out of date since the specifiation now says that the border shorthand sets all the
longhands of the the border-image shorthand to their initial values.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Updated.

Canonical link: https://commits.webkit.org/257966@main
